File: src/ios/API/LiveUpdateManager.swift
/**
 *   © Copyright 2016 IBM Corp.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

//
//  LiveUpdateManager.swift
//  A manager class for the  LiveUpdate APIs
//
//  Created by Oleg Sternberg & Ishai Borovoy on 14/1/16.
//  Copyright © 2016 IBM. All rights reserved.
//

import Foundation
import IBMMobileFirstPlatformFoundation
fileprivate func < <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l < r
  case (nil, _?):
    return true
  default:
    return false
  }
}

fileprivate func > <T : Comparable>(lhs: T?, rhs: T?) -> Bool {
  switch (lhs, rhs) {
  case let (l?, r?):
    return l > r
  default:
    return rhs < lhs
  }
}


open class LiveUpdateManager {
    fileprivate let bundleID : String = Bundle.main.bundleIdentifier!
    fileprivate let liveupdateClientScope : String = "liveupdate.mobileclient"
    fileprivate var applicationRoute: String = "";
    fileprivate var serviceURL: String = ""
    
    public static let sharedInstance = LiveUpdateManager()
    
    fileprivate init() {
        if let url = WLClient.init().serverUrl() {
            applicationRoute = url.absoluteString.replacingOccurrences(of: url.relativePath, with: "")
            serviceURL = "/mfpliveupdate/v1/\(bundleID)/configuration"
    }
    }
    
    /**
     Obtains a configuration from server / cache by a segment id
     
     - Parameter segment - the segment id which will be used by configuration adapter to return configuration
     
     - Parameter useCache - default is true, use false to explicitly obtain the configuration from the configuration adapter
     
     - Parameter completionHandler - the competition for retrieving the Configuration
     */
    open func obtainConfiguration (_ segment: String!, useCache: Bool = true, completionHandler: @escaping (_ configuration: Configuration?, _ error: NSError?) -> Void) {
        let encodedSegment = ecodeString(segment)
        let urlString = applicationRoute +  serviceURL + "/\(encodedSegment!)"
        if isValidUrl(url: urlString) {
            let url = URL(string: urlString)!
        OCLogger.getLogger().logDebugWithMessages("obtainConfiguration: segment = \(String(describing: segment)), useCache = \(useCache), url = \(url)")
        self.obtainConfiguration(segment, url: url, params: nil, useCache: useCache, completionHandler: completionHandler)
        } else {
            let error = "Failed to initialize Liveupdate instance because server url is nil. Check the server details in mfpclient.plist"
            OCLogger.getLogger().logDebugWithMessages(error)
            completionHandler(nil, NSError(domain: "", code: 0, userInfo: [NSLocalizedDescriptionKey : error]))
    }
    }
    
    /**
    Obtains a configuration from server / cache by params
     
     - Parameter params - the parameters which will be used by configuration adapter to return configuration
     
     - Parameter useCache - default is true, use false to explicitly obtain the configuration from the configuration adapter
     
     - Parameter completionHandler - the competition for retrieving the Configuration
     */
    open func obtainConfiguration (_ params: [String:String], useCache: Bool = true, completionHandler: @escaping (_ configuration: Configuration?, _ error: NSError?) -> Void) {
        let urlString = applicationRoute + bundleID + serviceURL
        if isValidUrl(url: urlString) {
            let url = URL(string: urlString)!
        let id = buildIDFromParams(params)
        OCLogger.getLogger().logDebugWithMessages("obtainConfiguration: params = \(params), useCache = \(useCache), url = \(url)")
        self.obtainConfiguration(id, url: url, params: params, useCache: useCache, completionHandler: completionHandler)
        } else {
            let error = "Failed to initialize Liveupdate instance because server url is nil. Check the server details in mfpclient.plist"
            OCLogger.getLogger().logDebugWithMessages(error)
            completionHandler(nil, NSError(domain: "", code: 0, userInfo: [NSLocalizedDescriptionKey : error]))
        }
    }
    
    
    fileprivate func obtainConfiguration (_ id : String, url: URL, params: [String: String]?, useCache: Bool, completionHandler: @escaping (_ configuration: Configuration?, _ error: NSError?) -> Void) {
        if let cachedConfig = LocalCache.getConfiguration(id) , useCache == true {
            // Get cached configuration
            OCLogger.getLogger().logDebugWithMessages("obtainConfiguration: Retrieved cached configuration. configuration = \(cachedConfig)")
            completionHandler(cachedConfig, nil)
        } else {
            sendConfigRequest(id, url: url, params: params) { configuration, error in
                OCLogger.getLogger().logDebugWithMessages("obtainConfiguration: Retrieving new configuration from server. configuration = \(String(describing: configuration))")
                completionHandler(configuration, error)
            }
        }
    }
    
    fileprivate func sendConfigRequest(_ id: String, url:URL, params: [String: String]?, completionHandler: @escaping (Configuration?, NSError?) -> Void) {
        let configurationServiceRequest = WLResourceRequest (url: url, method: WLHttpMethodGet, scope: liveupdateClientScope)
        
        OCLogger.getLogger().logTraceWithMessages("sendConfigRequest: id = \(id), url = \(url), params = \(String(describing: params))")
        
        if params != nil {
            for (paramName, paramValue) in params! {
                configurationServiceRequest?.setQueryParameterValue(paramValue, forName: paramName)
            }
        }
        configurationServiceRequest?.send { wlResponse, wlError in
            var configuration: Configuration? = nil
            
            if (wlError == nil) {
                var json = wlResponse?.responseJSON as? [String: AnyObject]
                
                if json == nil {
                    OCLogger.getLogger().logFatalWithMessages("sendConfigRequest: invalid JSON response")
                    json = [String: AnyObject]()
                }
                configuration = ConfigurationInstance(id :id, data: json!)
                // Save to cache
                
                OCLogger.getLogger().logTraceWithMessages("sendConfigRequest: saving configuration to cache. configuration = \(String(describing: configuration))")
                LocalCache.saveConfiguration(configuration!)
            } else {
                OCLogger.getLogger().logFatalWithMessages("sendConfigRequest: error while retriving configuration from server. error = \(String(describing: wlError))")
            }
            completionHandler(configuration, wlError as NSError?)
        }
    }
    
    fileprivate func buildIDFromParams (_ params: [String: String]?)->String {
        OCLogger.getLogger().logTraceWithMessages("buildIDFromParams: params = \(String(describing: params))")
        var paramsId = ""
        if (params?.count > 0) {
            for (paramName, paramValue) in params! {
                paramsId += "_\(paramName)_\(paramValue)"
            }
        }
        OCLogger.getLogger().logTraceWithMessages("buildIDFromParams: paramsId = \(paramsId)")
        return paramsId
    }
    
    fileprivate func ecodeString(_ path: String?) -> String? {
        return path?.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlHostAllowed)
    }
    
    fileprivate func isValidUrl(url: String) -> Bool {
        let urlRegEx = "http[s]?://(([^/:.[:space:]]+(.[^/:.[:space:]]+)*)|([0-9](.[0-9]{3})))(:[0-9]+)?((/[^?#[:space:]]+)([^#[:space:]]+)?(#.+)?)?"
        let urlTest = NSPredicate(format:"SELF MATCHES %@", urlRegEx)
        let result = urlTest.evaluate(with: url)
        return result
    }
}

File: src/android/com/worklight/ibmmobilefirstplatformfoundationliveupdate/LiveUpdateManager.java
/**
 *   © Copyright 2016 IBM Corp.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
package com.worklight.ibmmobilefirstplatformfoundationliveupdate;

import android.content.Context;
import android.webkit.URLUtil;

import com.worklight.common.Logger;
import com.worklight.ibmmobilefirstplatformfoundationliveupdate.api.Configuration;
import com.worklight.ibmmobilefirstplatformfoundationliveupdate.api.ConfigurationListener;
import com.worklight.ibmmobilefirstplatformfoundationliveupdate.cache.LocalCache;
import com.worklight.wlclient.api.WLClient;
import com.worklight.wlclient.api.WLFailResponse;
import com.worklight.wlclient.api.WLResourceRequest;
import com.worklight.wlclient.api.WLResponse;
import com.worklight.wlclient.api.WLResponseListener;

import org.json.JSONObject;

import java.net.URI;
import java.net.URL;
import java.util.Map;

/**
 * LiveUpdateManager
 * </p>
 * A manager class for the  LiveUpdate APIs
 *
 * @author Ishai Borovoy
 * @since 8.0.0
 */

public class LiveUpdateManager {

    private static LiveUpdateManager instance = null;
    private final static String LIVEUPDATE_CLIENT_SCOPE = "liveupdate.mobileclient";
    private String SERVICE_URL;


    private static final Logger logger = Logger.getInstance(LiveUpdateManager.class.getName());
    /**
     * getInstance
     *
     * @param context Android {@link Context}
     * @return LiveUpdateManager singleton instance
     */
    public static synchronized LiveUpdateManager getInstance(Context context) {
        if (instance == null) {
            instance = new LiveUpdateManager(context);
        }
        return instance;
    }

    private LiveUpdateManager(Context context) {
        try {
            WLClient client = null;
            // Get the applicationId and backend route from core
            try {
                client = WLClient.getInstance();
            } catch (Exception e) {
                client = WLClient.createInstance(context);
            }
            URL url = client.getServerUrl();
            StringBuilder routeBuilder = new StringBuilder();
            routeBuilder.append(url.getProtocol());
            routeBuilder.append("://");
            routeBuilder.append(url.getHost());
            int port = url.getPort();
            if(port != -1) {
                routeBuilder.append(":");
                routeBuilder.append(port);
            }
            String urlPath = url.getPath();
            String[] split = urlPath.split("/api");
            String contextRoute = urlPath.substring(0, split[0].lastIndexOf( "/" ));
            routeBuilder.append(contextRoute);

            String applicationRoute = routeBuilder.toString();
            String packageName = context.getPackageName();
            SERVICE_URL =  applicationRoute + "/mfpliveupdate/v1/" + packageName + "/configuration";
            if (!URLUtil.isValidUrl(SERVICE_URL)) {
                throw new RuntimeException("LiveUpdateManager:initialize() - An error occured while initializing Liveupdate service. Reason : Invalid HOST URL");
            }
        } catch (Exception e) {
            logger.error("LiveUpdateManager:initialize() - An error occured while initializing Liveupdate service.");
            throw new RuntimeException(e);
        }
    }

    /**
     * obtainConfiguration - obtains a configuration from server / cache by a segment id
     * </p>
     * The cache is enabled for this API
     *
     * @param segmentId - the segment id
     * @param configurationListener - the configuration listener for receiving the configuration
     */
    public void obtainConfiguration (String segmentId, ConfigurationListener configurationListener) {
        this.obtainConfiguration(segmentId, true, configurationListener);
    }

    /**
     * obtainConfiguration - obtains a configuration from server / cache by params
     * </p>
     * The cache is enabled for this API
     *
     * @param params - the params used by the server to return a configuration.
     * @param configurationListener - the configuration listener for receiving the configuration
     */
    public void obtainConfiguration (Map<String,String> params, ConfigurationListener configurationListener) {
        this.obtainConfiguration(params, true, configurationListener);
    }


    /**
     * obtainConfiguration - obtains a configuration from server / cache
     * @param useCache - true to use cache, false to always obtains from server
     * @param configurationListener - the configuration listener for receiving the configuration
     */
    public void obtainConfiguration (boolean useCache, ConfigurationListener configurationListener) {
        URI url = URI.create(SERVICE_URL + "/all");


        logger.debug("obtainConfiguration: useCache = " + useCache + ", url = " + url);
        this.obtainConfiguration("all", url, null, useCache, configurationListener);
    }

    /**
     * obtainConfiguration - obtains a configuration from server / cache by a segment id
     * @param segmentId - the segment id
     * @param useCache - true to use cache, false to always obtains from server
     * @param configurationListener - the configuration listener for receiving the configuration
     */
    public void obtainConfiguration (String segmentId, boolean useCache, ConfigurationListener configurationListener) {
        URI url = URI.create(SERVICE_URL + "/" + segmentId);


        logger.debug("obtainConfiguration: segment = " + segmentId + ", useCache = " + useCache + ", url = " + url);
        this.obtainConfiguration(segmentId, url, null, useCache, configurationListener);
    }

    /**
     * obtainConfiguration - obtains a configuration from server / cache by params
     * @param params - the params used by the server to return a configuration
     * @param useCache - true to use cache, false to always obtain configuration from server
     * @param configurationListener - the configuration listener for receiving the configuration
     */
    public void obtainConfiguration (Map<String,String> params, boolean useCache, ConfigurationListener configurationListener) {
        URI url = URI.create(SERVICE_URL);
        String id = buildIDFromParams(params);

        logger.debug("obtainConfiguration: params = " + params + ", useCache = " + useCache + ", url = " + url);
        this.obtainConfiguration(id, url, params, useCache, configurationListener);
    }


    private void obtainConfiguration (String id, URI url, Map<String,String> params, boolean useCache, final ConfigurationListener configurationListener) {
        Configuration cachedConfiguration = LocalCache.getConfiguration(id);

        if (cachedConfiguration != null && useCache) {
            logger.debug("obtainConfiguration: Retrieved cached configuration. configuration = " + cachedConfiguration);
            configurationListener.onSuccess(cachedConfiguration);
        } else {
            sendConfigRequest(id, url, params, configurationListener);
        }
    }


    private void sendConfigRequest(final String id, URI url, Map<String,String> params, final ConfigurationListener configurationListener) {
        WLResourceRequest configurationServiceRequest = new WLResourceRequest(url, WLResourceRequest.GET, LIVEUPDATE_CLIENT_SCOPE);

        logger.trace("sendConfigRequest: id = " + id + ", url = " + url + "params = " + params);

        if (params != null) {
            for (String paramKey : params.keySet()) {
                configurationServiceRequest.setQueryParameter(paramKey, params.get(paramKey));
            }

        }

        configurationServiceRequest.send(new WLResponseListener() {
            @Override
            public void onSuccess(WLResponse wlResponse) {
                JSONObject json = wlResponse.getResponseJSON();

                if (json == null) {
                    logger.error("sendConfigRequest: invalid JSON response");
                    json = new JSONObject();
                }
                Configuration configuration = new ConfigurationInstance(id, json);
                // Save to cache

                logger.trace("sendConfigRequest: saving configuration to cache. configuration = " +configuration);
                LocalCache.saveConfiguration(configuration);
                configurationListener.onSuccess(configuration);
            }

            @Override
            public void onFailure(WLFailResponse wlFailResponse) {
                logger.error("sendConfigRequest: error while retriving configuration from server. error = " + wlFailResponse.getErrorMsg());
                configurationListener.onFailure(wlFailResponse);
            }
        });
    }

    private String buildIDFromParams (Map<String,String> params) {
        logger.trace("buildIDFromParams: params = " + params);
        String paramsId = "";
        if (params != null && params.size() > 0) {
            for (String paramKey : params.keySet()) {
                paramsId += "_" + paramKey + "" + params.get(paramKey);
            }
        }
        logger.trace("buildIDFromParams: paramsId = " + paramsId);
        return paramsId;
    }
}
